### PR TITLE
Remove redundant product validity check from ConversionTrackProducer and modernize it

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR is part of https://github.com/cms-sw/cmssw/issues/30074 even though it itself does not fix any workflow. The bare "throw" will lead to abort (*) because there is no exception to be thrown. On the other hand, the `beamSpotHandle->position() `will throw an exception if the product is missing (with much clearer message), so this PR proposes just remove the check.

(*) https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/raw/slc7_amd64_gcc820/CMSSW_11_2_X_2020-06-01-2300/pyRelValMatrixLogs/run/135.9_ZMM_13+ZMMFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZMM_13+ZMMFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

In addition this PR modernizes `ConversionTrackProducer` by
* Constructing BeamSpot position as `math::XYZVector` once
* Using `edm::Event::get()` instead of going via `edm::Handle`
* Consuming trajectory maps only if `useTrajectory == True`
* Using `edm::ESGetToken`
* Removing unnecessary member variables

#### PR validation:

Workflow 135.9 does not abort anymore (but it won't succeed either, the real fix is different but independent of this PR). Limited matrix runs.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: RecoEgamma/EgammaPhotonProducers/src/ConversionTrackProducer.cc [hunk 1]
```diff
@@ -101,11 +101,6 @@ void ConversionTrackProducer::produce(edm::Event& e, const edm::EventSetup& es)
   edm::ESHandle<MagneticField> magFieldHandle;
   es.get<IdealMagneticFieldRecord>().get(magFieldHandle);
 
-  if (filterOnConvTrackHyp && !beamSpotHandle.isValid()) {
-    edm::LogError("Invalid Collection") << "invalid collection for the BeamSpot";
-    throw;
-  }
```
Comment (makortel) [inline RecoEgamma/EgammaPhotonProducers/src/ConversionTrackProducer.cc:107]: The `beamspotHandle->position()` on line 114/109 will throw an exception (with much clearer message) if the Handle is not valid.

(by the way, does
```cpp
math::XYZVector beamSpot = math::XYZVector(beamSpotHandle->position());
```
really need to be repeated for each input track?)

**Comment by perrotta** [comment]
May I suggest then to leave the LogError, without throwing? Perhaps, while adding to the report also the ```beamSpotInputTag```, so that one can have a complete picture from the log about who's the caller, and which is the missing beamSpot? (By the way: what is the extra info conveyed by ```beamspotHandle->position()```?)
Moreover, as you correctly pointed out, the beam spot doesn't need to be re-defined for every track: we can profit of this PR to take it out of the ```for``` loop, then.

**Comment by makortel** [comment]
> May I suggest then to leave the LogError, without throwing? Perhaps, while adding to the report also the `beamSpotInputTag`, so that one can have a complete picture from the log about who's the caller, and which is the missing beamSpot? (By the way: what is the extra info conveyed by `beamspotHandle->position()`?)

Why? If `beamSpotHandle` is not valid, `beamSpotHandle->position()` will throw the standard `ProductNotFound` exception that shows who tried to access which product.

**Comment by perrotta** [comment]
@makortel : Ok about removing the LogError as a whole from the ConversionTrackProducer
Please let me know if you intend to profit of this PR to also move out of the for loop the access to the beamSpot, as you already suggested in https://github.com/cms-sw/cmssw/pull/30082#discussion_r434104157

**Comment by makortel** [comment]
> Please let me know if you intend to profit of this PR to also move out of the for loop the access to the beamSpot

I can do that.

**Comment by makortel** [comment]
@perrotta Done, I could not resists on doing a bit more modernization on the same go. I hope it is fine.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by perrotta** [comment]
> @perrotta Done, I could not resists on doing a bit more modernization on the same go. I hope it is fine.

Great! Thank you @makortel

**Comment by perrotta** [comment]
+1
- Technical
- Jenkins tests pass

**Comment by silviodonato** [comment]
+1

**Pull request merged**